code-experiments/src/brentq.c

```diff
@@ -0,0 +1,195 @@
+    - make it standalone
+      - removing SciPy-dependent stuff
+      - merging with zeros.h declaration file
+    - wrapping in brentinv(f, y, ...)
+      - automatic range interval computation [xa, xb]
+      - assume f has fixed point in 0
+ */
+
+#include <math.h>
+
+typedef double (*callback_type)(double, void*);
+
+static double brentq(callback_type f, double y, double xa, double xb, double xtol,
+                     double rtol, int iter, void *func_data);
+
+static double brentinv(callback_type f, double y, void *func_data);
+
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+
+/*
+  Note: signbit is in the C99 math library and we compile with C89 standard.
+*/
+#define signbit(x)((x) < 0 ? 1 : 0)
+
+/*
+  At the top of the loop the situation is the following:
+
+    1. the root is bracketed between xa and xb
+    2. xa is the most recent estimate
+    3. xp is the previous estimate
+    4. |fp| < |fb|
+
+  The order of xa and xp doesn't matter, but assume xp < xb. Then xa lies to
+  the right of xp and the assumption is that xa is increasing towards the root.
+  In this situation we will attempt quadratic extrapolation as long as the
+  condition
+
+  *  |fa| < |fp| < |fb|
+
+  is satisfied. That is, the function value is decreasing as we go along.
+  Note the 4 above implies that the right inequality already holds.
+
+  The first check is that xa is still to the left of the root. If not, xb is
+  replaced by xp and the interval reverses, with xb < xa. In this situation
+  we will try linear interpolation. That this has happened is signaled by the
+  equality xb == xp;
+
+  The second check is that |fa| < |fb|. If this is not the case, we swap
+  xa and xb and resort to bisection.
+
+*/
+
+double brentq(callback_type f, double y, double xa, double xb, double xtol, double rtol,
+       int iter, void *func_data)  {
+
+    double xpre = xa, xcur = xb;
+    double xblk = 0., fpre, fcur, fblk = 0., spre = 0., scur = 0., sbis;
+    /* the tolerance is 2*delta */
+    double delta;
+    double stry, dpre, dblk;
+    int i;
+
+    fpre = (*f)(xpre, func_data) - y;
+    fcur = (*f)(xcur, func_data) - y;
+    if (fpre*fcur > 0) {
+        /* Sign error */
+        return NAN;
+    }
+    if (fpre == 0) {
+        /* Converged */
+        return xpre;
+    }
+    if (fcur == 0) {
+        /* Converged*/
+        return xcur;
+    }
+
+    for (i = 0; i < iter; i++) {
+        if (fpre != 0 && fcur != 0 &&
+	    (signbit(fpre) != signbit(fcur))) {
+            xblk = xpre;
+            fblk = fpre;
+            spre = scur = xcur - xpre;
+        }
+        if (fabs(fblk) < fabs(fcur)) {
+            xpre = xcur;
+            xcur = xblk;
+            xblk = xpre;
+
+            fpre = fcur;
+            fcur = fblk;
+            fblk = fpre;
+        }
+
+        delta = (xtol + rtol*fabs(xcur))/2;
+        sbis = (xblk - xcur)/2;
+        if (fcur == 0 || fabs(sbis) < delta) {
+            /* TODO: converged*/
```

Comment:
Fix todo?

Reply:
Yes, just forgotten, nothing special to do here